include/small/obuf_asan.h

```diff
@@ -0,0 +1,200 @@
+	/**
+	 * Number of vectors at the beginning of iovec vector list
+	 * that are allocated with distinct malloc call. Let's refer
+	 * them as checked vectors.
+	 */
+	SMALL_OBUF_IOV_CHECKED_SIZE = SMALL_OBUF_IOV_MAX + 1 -
+				      SMALL_OBUF_IOV_GEOMETRIC_SIZE,
+	/** Alignment for the checked vectors. */
+	SMALL_OBUF_ALIGNMENT = 1,
+	/** Minimum size when reserving memory. See struct obuf comment. */
+	SMALL_OBUF_MIN_RESERVE = 128,
+};
+
+/**
+ * ASAN friendly implementation for obuf allocator. It has same interface as
+ * regular implementation but at the beginning it uses malloc() for every
+ * allocation. More precisely for the first (in terms of obuf_iovcnt())
+ * SMALL_OBUF_IOV_CHECKED_SIZE allocations. Then the memory is allocated in
+ * exponentially growing blocks which then used to provide requested
+ * allocations as in the regular implementation.
+ *
+ * This combined strategy allows to do out-of-bound access checks for first
+ * allocations on the one hand and on the other hand limit the size of iov
+ * vector list. The latter is required to avoid iov vector list reallocations
+ * which is expected by some client code.
+ *
+ * See however a bit of limitation for out-of-bound access check in description
+ * of small_asan_alloc.
+ *
+ * Allocations are completely not aligned in the beginning (that is not
+ * 2 aligned). This improves unaligned memory access check.
+ *
+ * Also at the beginning we reserve at least SMALL_OBUF_MIN_RESERVE bytes
+ * so that runtime code path with ASAN implementation is similar to regular
+ * one. The overreserve is not very large comparing to regular one to reduce
+ * memory waste.
+ */
+struct obuf {
+	/** For compatibility with existing code only. */
+	struct slab_cache *slabc;
+	/**
+	 * List of iovec vector. NULL terminated. Allocated at creation
+	 * and does not reallocated. Size is SMALL_OBUF_IOV_MAX + 1.
+	 */
+	struct iovec *iov;
+	/**
+	 * Vectors starting from SMALL_OBUF_IOV_CHECKED_SIZE index are
+	 * exponentially growing with factor 2 starting from this size.
+	 * Note that size can be larger than the described if larger allocation
+	 * is requested.
+	 */
+	size_t start_capacity;
+	/**
+	 * Capacity of exponentially growing vectors. Note that size is off
+	 * by one because iov list is NULL terminated.
+	 */
+	size_t capacity[SMALL_OBUF_IOV_GEOMETRIC_SIZE - 1];
+	/**
+	 * If pos == 0 and iov[0] == NULL then obuf is empty. Otherwise
+	 * pos is index of the vector in the vectors list holding last
+	 * allocation. A bit odd but follows regular implementation
+	 * semantics for compatibility.
+	 */
+	int pos;
+	/** Total size of allocations. */
+	size_t used;
+	/** If not 0 then amount of memory reserved with the obuf_reserve. */
+	size_t reserved;
+};
+
+void
+obuf_create(struct obuf *buf, struct slab_cache *slabc, size_t start_capacity);
+
+static inline bool
+obuf_is_initialized(const struct obuf *buf)
+{
+	return buf->iov != NULL;
+}
+
+static inline size_t
+obuf_size(struct obuf *obuf)
```

Comment:
Nit: Please use the same name for the local variable denoting `obuf`.

Reply:
Fixed.

```diff
--- a/include/small/obuf_asan.h
+++ b/include/small/obuf_asan.h
@@ -132,9 +132,9 @@ obuf_is_initialized(const struct obuf *buf)
 }

 static inline size_t
-obuf_size(struct obuf *obuf)
+obuf_size(struct obuf *buf)
 {
-       return obuf->used;
+       return buf->used;
 }
```